### ISSUE TITLE
Proposal: Storing comments in keys

### ISSUE THREAD

**Issue by homebysix** [ProfileCreator/ProfileManifests#567]
There's been some [discussion](https://macadmins.slack.com/archives/CH9LT65U6/p1658690347034599) [lately](https://macadmins.slack.com/archives/CH9LT65U6/p1658782066948909) about performing some programmatic comparisons and/or updates of the keys in this repository against the keys in Apple's [device-management](https://github.com/apple/device-management) repository. However, neither `plutil` nor Python's `plistlib` handles `<!-- HTML/XML-style comments -->` like the ones currently used in many manifests in this repo, making programmatic comparison difficult without losing important context.

My proposal is that we choose a naming convention for comments, and then promote existing comments to their own key in order to preserve them during programmatic operations. Two possible approaches to consider include:

a. Per-manifest comment using a specific key. @relgit proposed something like `pfm_x_comment` which would reduce the possibility of a naming collision if the PFM specs get expanded someday. The major downside to this approach is that [some manifests](https://github.com/ProfileCreator/ProfileManifests/blob/master/Manifests/ManagedPreferencesApplications/com.github.macadmins.Nudge.plist#L217) are complex enough that keeping comments nearby specific keys might be beneficial.

b. Per-key comment using a specific suffix, like `<key_name>_comment`, which would allow the comment to remain sorted adjacent to the key it's describing. The downside to this approach is that reserving a suffix's namespace is a much greater long term commitment than simply reserving one specific comment key name. There may be collisions, and the convention would be harder to enforce.

Something not addressed by either option above: comments that live in the space between array items, like [this one](https://github.com/ProfileCreator/ProfileManifests/blob/master/Manifests/ManagedPreferencesApplications/com.github.macadmins.Nudge.plist#L217). Ideally, in extremely complex cases, we could point to the app's documentation itself for a detailed guide to the keys, rather than relying on inline comments at all.

Interested to hear community ideas about this, because converting comments to data keys is step one towards many potential automations of this repo in the future.

**Comment by relgit** [comment]
I took a moment to scan all of the manifests in `Manifests` (apart for the developer ones) for comments and to better understand the scope of this problem, and I came up with a number of insights:

* There are only 26 relevant manifests with comments.
* Most comments are used to group together subkeys. Very few actually hold per-subkey information, and in 7 manifests comments are used to hide draft work.
* Most, if not all, of the comments are useful only to contributors on the repository, and not to end users.

Now, looking at this makes me think that we can and should just do away with most of the comments altogether. Here's why.

For starters, I feel that comments used for grouping are not necessary. Throughout my work with the manifests, I can't recall a time that I personally relied on them. Perhaps others do, but is the benefit gained from them worth the effort that will be required to keep them?

Next, draft work should not exist in comments. It should either be in draft branches, or should be finalized and made available to users. For example in the [`com.agilebits.onepassword7` manifest](https://github.com/ProfileCreator/ProfileManifests/blob/master/Manifests/ManagedPreferencesApplications/com.agilebits.onepassword7.plist#L561) there's a key commented due to a bug in 1Password. Assuming this is still ongoing, if the subkey is documented it should be available to users and up to them to decide on using it or not, with perhaps a small `pfm_note` message to inform about the bug.

Lastly, in the few cases where comments are actually relevant for specific subkeys, for example on line 315 in the [`com.apple.dock` manifest](https://github.com/ProfileCreator/ProfileManifests/blob/master/Manifests/ManifestsApple/com.apple.dock.plist#L315), we can definitely convert them to a `pfm_x_comment` style key-value pair and keep them in.

I think that if we agree and put this plan in motion, we'd be much quicker to acheive `plutil`-readiness in the repository.

What do you think?

**Comment by apizz** [comment]
I appreciate the thinking here and agree that a key would be beneficial.

I don't see why we couldn't have a `pfm_x_comment` that we we can use at the top-most level of the manifest and within preference dictionaries.  This is similar to how `pfm_documentation_url` works.

While I agree that there may be some situations where `pfm_x_comment` would benefit from being a bit more unique and closer to the item it's commenting on, I think for 95%+ situations this wouldn't require doing a bunch of suffixing.  Even where we're trying to make it clearer the start / end of certain sections, I would think a simple dictionary would suffice. Example:

```
<dict>
    <key>pfm_x_comment</key>
    <string>This is a comment RE the somePreference key</string>
    <key>pfm_name</key>
    <string>somePreference</string>
    <key>pfm_type</key>
    <string>string</string>
</dict>
<dict>
    <key>pfm_x_comment</key>
    <string>this is a comment about the above section of preferences</string>
</dict>
```

I'll get something going in the existing manifests to get your feedback

**Comment by relgit** [comment]
@apizz sure root-level comment tags are fine with me, but it's going to be a nay from me on the non-key dictionaries option ☺️

I think those have the potential to break some consuming software because they miss the expected `pfm_name` and `pfm_type`, and personally I find that they're only useful for grouping-comments - which I suggest we axe.

**Comment by relgit** [comment]
Addressed in #609.
